File: TokenMihariban/Logic/SessionBlockCalculator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TokenMihariban.Models;

namespace TokenMihariban.Logic;

/// <summary>
/// Groups usage events into estimated 5-hour session blocks — same heuristic as every
/// other platform (community "ccusage"-style): a block starts at the top of the hour of
/// the first event after either a full 5 hours elapsed or a 5+ hour activity gap.
/// Anthropic doesn't publish the real algorithm; treat this as an estimate.
/// </summary>
public static class SessionBlockCalculator
{
    public static readonly TimeSpan BlockDuration = TimeSpan.FromHours(5);

    public static List<SessionBlock> ComputeBlocks(IReadOnlyList<UsageEvent> events)
    {
        var blocks = new List<SessionBlock>();
        if (events.Count == 0) return blocks;

        var sorted = events.OrderBy(e => e.Timestamp).ToList();

        DateTime? currentStart = null;
        var currentEvents = new List<UsageEvent>();

        void Flush()
        {
            if (currentStart is not { } start || currentEvents.Count == 0) return;
            blocks.Add(new SessionBlock(start, start + BlockDuration, currentEvents.ToList()));
        }

        foreach (var evt in sorted)
        {
            if (currentStart is not { } start)
            {
                currentStart = FlooredToHourUtc(evt.Timestamp);
                currentEvents = new List<UsageEvent> { evt };
                continue;
            }

            var blockEnd = start + BlockDuration;
            var lastEventTime = currentEvents.Count > 0 ? currentEvents[^1].Timestamp : start;
            var gapSinceLastEvent = evt.Timestamp - lastEventTime;

            if (evt.Timestamp >= blockEnd || gapSinceLastEvent >= BlockDuration)
            {
                Flush();
                currentStart = FlooredToHourUtc(evt.Timestamp);
                currentEvents = new List<UsageEvent> { evt };
            }
            else
            {
                currentEvents.Add(evt);
            }
        }
        Flush();

        return blocks;
    }

    /// <summary>
    /// Matches the Mac/iOS/Android apps' explicit UTC flooring — block boundaries must
    /// line up across platforms and show the same
    /// current block/reset time regardless of which device's clock/timezone computed it.
    /// </summary>
    private static DateTime FlooredToHourUtc(DateTime date)
    {
        var utc = date.Kind == DateTimeKind.Utc ? date : date.ToUniversalTime();
        return new DateTime(utc.Year, utc.Month, utc.Day, utc.Hour, 0, 0, DateTimeKind.Utc);
    }
}

File: macOS/Packages/ClaudeUsageCore/Sources/ClaudeUsageCore/SessionBlockCalculator.swift
import Foundation

/// Groups usage events into estimated 5-hour session blocks.
///
/// This mirrors the heuristic used by community tools (e.g. ccusage) since Anthropic
/// does not publish how the Pro/Max 5-hour window is actually computed: a new block
/// starts at the top of the hour of the first event after either (a) the previous
/// block ran its full 5 hours, or (b) there was a 5+ hour gap with no activity.
/// Treat the result as an estimate, not an authoritative value.
public enum SessionBlockCalculator {
    public static let blockDuration: TimeInterval = 5 * 3600

    public static func computeBlocks(from events: [UsageEvent]) -> [SessionBlock] {
        guard !events.isEmpty else { return [] }
        let sorted = events.sorted { $0.timestamp < $1.timestamp }

        var blocks: [SessionBlock] = []
        var currentStart: Date?
        var currentEvents: [UsageEvent] = []

        func flush() {
            guard let start = currentStart, !currentEvents.isEmpty else { return }
            blocks.append(SessionBlock(start: start, end: start.addingTimeInterval(blockDuration), events: currentEvents))
        }

        for event in sorted {
            guard let start = currentStart else {
                currentStart = flooredToHour(event.timestamp)
                currentEvents = [event]
                continue
            }

            let blockEnd = start.addingTimeInterval(blockDuration)
            let gapSinceLastEvent = event.timestamp.timeIntervalSince(currentEvents.last?.timestamp ?? start)

            if event.timestamp >= blockEnd || gapSinceLastEvent >= blockDuration {
                flush()
                currentStart = flooredToHour(event.timestamp)
                currentEvents = [event]
            } else {
                currentEvents.append(event)
            }
        }
        flush()

        return blocks
    }

    /// The block still open at `referenceDate`, if any.
    public static func activeBlock(from events: [UsageEvent], at referenceDate: Date) -> SessionBlock? {
        guard let last = computeBlocks(from: events).last else { return nil }
        return referenceDate < last.end ? last : nil
    }

    private static func flooredToHour(_ date: Date) -> Date {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC") ?? .current
        let components = calendar.dateComponents([.year, .month, .day, .hour], from: date)
        return calendar.date(from: components) ?? date
    }
}
